File: heap/더맵게.py
import heapq
def check(h, k):
    for i in h:
        if i < k:
            return False
    return True

def solution(scoville, k):
    answer = 0
    h = []
    for i in scoville:
        heapq.heappush(h, i)
    
    while True:
        if len(h) == 1: 
            return -1
        first = heapq.heappop(h)
        second = heapq.heappop(h)
        heapq.heappush(h, first + second*2)
        answer += 1
        if check(h, k):
            return answer

File: heap/더맵게.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        PriorityQueue<Integer> queue = new PriorityQueue<>(Arrays.stream(scoville).boxed().collect(Collectors.toList()));
        
        while (queue.size() > 0) {
            int first = queue.poll();
            if (first >= K) return answer;
            if (queue.isEmpty()) {
                answer = -1;
                break;
            }
            int second = queue.poll();
            queue.offer(first + second * 2);
            answer++;
        }
        
        return answer;
    }
}
